feat/ant_linux_footpod

Bassai-Sho opened the pull request:
# Add ANT+ Virtual Footpod Broadcasting for Linux

## What This Feature Does

This PR adds ANT+ footpod broadcasting to QDomyos-Zwift on Linux. The application can now send treadmill speed and cadence data to Garmin watches and other ANT+ devices using an ANT+ footpod payload format compatible with Garmin devices.

**User benefit**: Connect your treadmill directly to your Garmin watch without needing a physical footpod sensor.

## Technical Implementation

### Architecture Overview

The feature uses a **hybrid C++/Python design**:

- **C++ Layer** (`AntManager`, `AntWorker`): Manages the ANT+ worker thread and bridges data from the treadmill to Python
- **Python Layer** (`ant_broadcaster.py`): Handles low-level USB communication with the ANT+ dongle using the `openant` library
- **Integration**: Python script is embedded in the binary using `pybind11` (no external files needed)

### Key Design Decisions

**Non-blocking delayed startup**: The ANT+ feature starts 10 seconds after device connection using `QTimer::singleShot`. This delay ensures the treadmill's blocking initialization (`btinit()`) completes before ANT+ broadcasting begins.

**No core file changes**: All new code is isolated in the `antlinux` directory and `main.cpp`. Existing device drivers remain unchanged for stability.

**Thread safety**: ANT+ broadcasting runs on a dedicated `QThread` to avoid blocking the main application.

### Initialization Flow

```
1. Application starts → creates bluetooth object
2. Bluetooth discovers treadmill → creates device object
3. Device starts blocking initialization (btinit, ~8 seconds)
4. After connection, main.cpp arms 10-second timer
5. Timer fires → starts ANT+ worker thread
6. Worker thread broadcasts treadmill data via ANT+ dongle
```

## Installation Requirements

### System Dependencies
```bash
sudo apt-get install -y libusb-1.0-0-dev libudev-dev python3-pip python3-venv
```

### Python Virtual Environment
Create a Python virtual environment (required for ANT+ libraries):
```bash
python3 -m venv ~/ant_venv
~/ant_venv/bin/python3 -m pip install pybind11 pyusb openant
```

**Note**: The application automatically finds this virtual environment at `~/ant_venv`.

### Hardware
- ANT+ USB dongle (tested with Garmin ANT+ USB-m, model 0fcf:1008/1009)
- Compatible ANT+ receiver (tested with Garmin Forerunner 245 Music)

## Usage

```bash
# Enable ANT+ footpod broadcasting
sudo ./qdomyos-zwift -no-gui -ant-footpod

# Use custom ANT+ device ID
sudo ./qdomyos-zwift -no-gui -ant-footpod -ant-device 12345

# Enable verbose ANT+ debug logging
sudo ./qdomyos-zwift -no-gui -ant-footpod -ant-verbose
```

## Files Changed

### New Files

**Core ANT+ Implementation:**
- `src/devices/antlinux/AntManager.cpp/h` - Singleton managing ANT+ feature lifecycle and worker thread creation
- `src/devices/antlinux/AntWorker.cpp/h` - Worker thread managing embedded Python interpreter and treadmill data feeding
- `src/devices/antlinux/ant_broadcaster.py` - Python script implementing ANT+ footpod protocol using openant library
- `src/devices/antlinux/ant_test_broadcaster.py` - Standalone test utility for verifying ANT+ dongle connectivity independently

**Build System:**
- `src/devices/antlinux/py_to_header.py` - Converts Python scripts to C++ header string literals for embedding
- `src/devices/antlinux/antlinux.pri` - Qt project configuration defining build flags and Python integration settings
- `src/devices/antlinux/build_check_list.sh` - Validates development environment setup including Python venv and USB permissions

**Documentation:**
- `src/devices/antlinux/README.md` - Complete installation guide
### Modified Files
- `src/main.cpp` - Added ANT+ command-line parsing and delayed startup trigger logic
- `src/devices/bluetooth.cpp` - Added device connection signal emission for ANT+ initialization
- `src/devices/bluetoothdevice.h` - Extended header with ANT+ integration declarations

## Testing

Tested configuration:
- **Platform**: Raspberry Pi Zero 2 W (Raspberry Pi OS Bookworm 64-bit)
- **Hardware**: Proform treadmill + Garmin ANT+ USB-m dongle
- **Receiver**: Garmin Forerunner 245 Music

## Known Limitations

1. **Linux only** - Currently supported only on Linux systems
2. **Python dependency** - Requires Python virtual environment setup with openant library
3. **Startup delay** - ANT+ broadcasting begins ~2 seconds after treadmill is fully ready (10-second total delay accounts for the ~8-second `btinit()` blocking initialization)

## Backward Compatibility

This feature is **completely optional** and does not affect existing functionality:

- Works alongside all existing Bluetooth features
- Disabled automatically if dependencies are missing
- All existing command-line arguments unchanged
- No modifications to core device driver code

If the ANT+ feature is not used or dependencies are unavailable, the application behaves exactly as before.

cagnulein commented:
Wow it looks amazing, thanks! I suggest to enable ANT_LINUX_ENABLED automatically in the pro file when building for Linux. What do you think?

Bassai-Sho commented:
That's a great suggestion. I completely agree that making the feature as easy as possible to enable is the ideal goal.

My only concern with enabling ANT_LINUX_ENABLED automatically for all Linux builds is that the ANT+ feature has a hard dependency on a specific Python virtual environment (ant_venv) and several development packages (python3-dev, pybind11, etc.). 

This hybrid C++/Python approach was chosen to leverage the mature and well-tested openant library, as a full native C++ implementation would have been a massive undertaking.

Because of these dependencies, if we define ANT_LINUX_ENABLED by default, the build would unfortunately fail for any user who hasn't manually performed this specific Python setup. This could also break the standard Linux CI build, as the runner won't have these dependencies installed.

To solve this, the current approach in antlinux.pri acts as an "opt-in" feature detector. It automatically checks for the existence of the .ant_venv_path file, which our setup scripts create as a signal that the environment is ready.

- If the file exists, the feature is seamlessly enabled.
- If it doesn't exist, it's safely skipped.

What do you think of this approach? It seems to provide the best balance, it offers "just-works" automation for users who follow the setup, while guaranteeing that the build remains safe and clean for everyone else and for the CI.

cagnulein commented:
Yes I like it! Also maybe we can try to adapt the raspberry images in the
CI to have the full setup ready? What do you think?

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno dom 19 ott 2025 alle 18:14 Adam Sharpe ***@***.***>
ha scritto:

> *Bassai-Sho* left a comment (cagnulein/qdomyos-zwift#3780)
> <https://github.com/cagnulein/qdomyos-zwift/pull/3780#issuecomment-3419777197>
>
> That's a great suggestion. I completely agree that making the feature as
> easy as possible to enable is the ideal goal.
>
> My only concern with enabling ANT_LINUX_ENABLED automatically for all
> Linux builds is that the ANT+ feature has a hard dependency on a specific
> Python virtual environment (ant_venv) and several development packages
> (python3-dev, pybind11, etc.).
>
> This hybrid C++/Python approach was chosen to leverage the mature and
> well-tested openant library, as a full native C++ implementation would have
> been a massive undertaking.
>
> Because of these dependencies, if we define ANT_LINUX_ENABLED by default,
> the build would unfortunately fail for any user who hasn't manually
> performed this specific Python setup. This could also break the standard
> Linux CI build, as the runner won't have these dependencies installed.
>
> To solve this, the current approach in antlinux.pri acts as an "opt-in"
> feature detector. It automatically checks for the existence of the
> .ant_venv_path file, which our setup scripts create as a signal that the
> environment is ready.
>
>    - If the file exists, the feature is seamlessly enabled.
>    - If it doesn't exist, it's safely skipped.
>
> What do you think of this approach? It seems to provide the best balance,
> it offers "just-works" automation for users who follow the setup, while
> guaranteeing that the build remains safe and clean for everyone else and
> for the CI.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cagnulein/qdomyos-zwift/pull/3780#issuecomment-3419777197>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALYWCUEVIVK3XRUR2GTVD3YO2GJAVCNFSM6AAAAACJTRF7ZSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTIMJZG43TOMJZG4>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

Bassai-Sho commented:
Hi Roberto,

I believe the ANT+ feature is now stable that I can start following the TESTPLAN.md. Before I do, I want to confirm the build plan with you.

For now, I have created two separate Docker builds for ANT+

- raspberry-pi-build-64bit-ant
- linux-x86-build-ant

This seemed like the safest approach. The original builds do not change, so there is zero risk for current users, and the new Python dependencies are kept separate.

My suggestion is to use these separate builds for testing and the first release. Later, once the feature is proven to be stable, we can make these the main builds for all Linux and Pi users.

Do you agree with this plan, or do you have a different preference?

Also, where is the best place for users to download these new builds? I could not find a "nightly builds" area. Right now, I can only find the download links deep inside the GitHub Actions workflow archives. Is there a better, more direct link?

Many Thanks,
Adam

Bassai-Sho committed:
Refinements and robustness updates

Bassai-Sho committed:
updates for ubuntu 24.04 default

Bassai-Sho committed:
Merge branch 'master' into feat/ant_footpod

Bassai-Sho committed:
Update setup-dashboard and fix newline for docker build

Bassai-Sho committed:
Merge branch 'master' into feat/ant_footpod

Bassai-Sho committed:
Resolve AntWorker virtual enviroment

Bassai-Sho committed:
Fix for python venv

Bassai-Sho committed:
QML Dependecy improvements that prevented GUI running

Bassai-Sho committed:
Merge master into feat/ant_footpod and resolve workflow conflicts

Bassai-Sho committed:
Fix linux bluetooth path and yaml rpi

Bassai-Sho committed:
virtual emulation conditions

Bassai-Sho committed:
Merge branch 'master' into feat/ant_footpod

Bassai-Sho committed:
docker and setup-dashboard updates

Bassai-Sho committed:
Merge branch 'master' into feat/ant_footpod

Bassai-Sho committed:
Remove workarounds to virtualtreadmill, fix now applied

Bassai-Sho committed:
Updated documentation

Bassai-Sho committed:
Improvements to service, ant+ monitor, added smart monitor

Bassai-Sho committed:
Added smart monitor and pi overlay menu options

Bassai-Sho committed:
Added ANT+ distance calibration

Bassai-Sho committed:
Updates to wizard and setup-dashboard

Bassai-Sho committed:
Merge branch 'master' into feat/ant_footpod

Bassai-Sho committed:
Calibration updates

Bassai-Sho committed:
Merge branch 'master' into feat/ant_footpod

Bassai-Sho committed:
Fix virtual treadmill FTMS control and ANT+ dashboard improvements

virtualtreadmill.cpp: advertise speed+incline target setting features
in 0x2ACC so FTMS clients send Set Target Speed/Inclination opcodes

characteristicwriteprocessor2ad9.cpp: forward Start/Stop opcodes to
real treadmill; guard Start on currentSpeed==0 to prevent re-init on
Resume; respect treadmill_force_speed for speed/incline commands;
honour FTMS pause parameter byte on Stop

setup-dashboard.sh: fix QZ Service status cell padding; fix
enable/disable targeting wrong service when monitor installed; pause
monitor on Stop to prevent immediate restart; add Force Treadmill Speed
toggle to QZ Runtime Flags menu

docker-build.sh: accept aarch64 as alias for arm64

Bassai-Sho committed:
Merge branch 'master' into feat/ant_footpod

Bassai-Sho committed:
Add ANT+ startup countdown toasts and fix SD card cell on non-Pi

- main.cpp: show countdown toasts (t=0, t=4s, t=7s) during the ANT+
  startup delay so the user knows why tiles are unresponsive; confirm
  with "ANT+ ready" or "ANT+ not started" on completion. All changes
  are within the existing ANT_LINUX_ENABLED block.
- setup-dashboard.sh: clear R2_key (not just R2_label) for overlay_fs
  on non-Pi so the SD Card Wear cell is fully suppressed and the grey
  dot no longer renders on desktop Linux.

Bassai-Sho committed:
Emit true ANT+ ready toast when broadcasting is confirmed active

Add broadcastingStarted() signal to AntWorker, emitted after Python
init and dongle reset complete. AntManager forwards it; main.cpp
connects with SingleShotConnection to show "ANT+ ready" toast only
once broadcasting is actually live (~3 s after startForDevice()).
The intermediate toast changes to "ANT+ starting…" so the user
sees accurate status throughout the startup sequence.

Bassai-Sho committed:
Merge branch 'master' into feat/ant_footpod

Bassai-Sho committed:
Make QML dep check dynamic: inject modules from source scan at build time

The Dockerfile now scans all .qml files after make and patches
REQUIRED_MODULES=() in check-qml-deps.sh with the exact imports used
by that build. This means adding a new QML screen automatically
updates the dependency check with zero manual maintenance.

When REQUIRED_MODULES is empty (unpatched/dev copy), the script falls
back to checking every entry in QML_MAP. Also adds Qt.labs.folderlistmodel
and fixes the check to scan the filesystem directly rather than using
strings(1) on the binary, which fails when Qt Quick Compiler is used.

Bassai-Sho committed:
Fix off-by-one in device generator and exclude feature-flag switches

parse_inline_model_array was filtering out 'Other' (index 0), shifting
every subsequent model item down by 1. The initializeModel() ternary
uses 1-based indices, so every device was assigned the setting key of
the previous model — e.g. 'Proform 705 CST' got proform_treadmill_9_0,
'Nordictrack x14i' got proform_treadmill_705_cst.

Fix: keep 'Other' in the array so loop indices align with the ternary.

Also adds EXCLUDED_SETTING_IDS to filter feature-flag IndicatorOnlySwitch
elements (confirm_stop_workout, sole/domyos elliptical inclination) that
were incorrectly appearing as device models in the Elliptical section.

Regenerated devices.ini, devices_optimized.json and .menu_cache.

Bassai-Sho committed:
Inject only extra QML modules not in QML_MAP; rename to ADDITIONAL_MODULES

- Renamed REQUIRED_MODULES to ADDITIONAL_MODULES — clearer that these
  are modules beyond what QML_MAP already covers
- Dockerfile injection now diffs scanned modules against QML_MAP keys
  using temp files (POSIX-compatible, avoids bash process substitution)
- When all scanned modules are covered, injects a comment confirming
  the scan ran cleanly rather than leaving an empty array unexplained
- Copies patched check-qml-deps.sh from builder stage (not source tree)
  so the injection is preserved in the artifact
- Adds debug output to injection step for visibility in build logs

Bassai-Sho committed:
Merge branch 'master' into feat/ant_footpod

Bassai-Sho committed:
Always recheck status after install regardless of install exit code

All install+recheck pairs used && which skipped the recheck if the
install returned non-zero (e.g. already installed, transient apt
warning). Changed to ; so the status check always runs and the
dashboard reflects the true post-install state immediately.

Affected: venv, python_packages, qt5_libs/qml_modules, bluetooth,
plugdev, udev_rules, lsusb.

Bassai-Sho committed:
Split deep clean: QML modules safe on desktop, core Qt5 libs headless-only

QML modules (qml-module-*) are not desktop dependencies and can safely
be removed on any system — useful for testing the detection/reinstall
flow on desktop Linux.

Core Qt5 libs (libqt5core5a, libqt5qml5, libqt5quick5 etc.) are shared
with GNOME/GDM and must only be purged on headless systems. On desktop
systems these are now skipped with a clear warning rather than blocking
the entire deep clean.

Bassai-Sho committed:
[ANT_LINUX_ENABLED] Standard FTMS mode for Pi virtual treadmill

Adds -no-wahoo-service flag (main.cpp) that switches the virtual treadmill
from KICKR RUN/Wahoo mode to standard FTMS ('Horizon 7.0 AT'), guarded by
ANT_LINUX_ENABLED throughout virtualtreadmill.cpp:

- BLE name, DIS strings (manufacturer/firmware) set to Horizon Fitness values
- 0x2ACC feature flags: 0x0D (avg-speed+distance+incline) and speed+incline
  target bits enabled for treadmill apps that gate CP commands on these
- CP (0x2AD9) CCCD pre-set to 0x02 (indications) so non-subscribing clients
  receive responses
- 0x2AD2 (Indoor Bike Data) omitted in standard mode to avoid confusing
  treadmill-only apps
- 0x2A00 GAP device name aligned to advertising name

setup-dashboard.sh: 'Standard FTMS Mode' toggle added to service config UI.

Bassai-Sho committed:
Merge branch 'master' into feat/ant_footpod

Bassai-Sho committed:
setup-dashboard: optimize Pi service config and monitor

- Cleaned up Pi service menu by removing unnecessary options.
- Renamed status labels (QZ Monitor, SD Overlay FS) for clarity.
- Increased Monitor responsiveness (20s polling) and reduced grace period to 5m.
- Suppressed redundant service restart prompts when Monitor is active.

Bassai-Sho commented:
Hi Roberto,

The ANT+ feature is now working on both the Raspberry Pi (no screen) and the Desktop (Ubuntu 24.04). I am just fixing a few things in the setup-dashboard.sh script so the user has the tools to setup ANT+ on Linux. I will then submit the changes for you to check.

**The "Runna" App Fix**

I also fixed a problem so that the Runna app now works with QZ on Pi.
-     Someone on Facebook asked if they worked using two device
-     It might work on iPhones, but I can't test that.
-     On the Raspberry Pi, it was not working

**What was the problem?**
See 391db60a (Raspberry kickr run, #2941
In December 2024, some code was added to use a "Wahoo" style connection. This was causing the Runna app to crash. I thought it was Runna's fault, but I found a fix for 'virtualtreadmill.cpp'. When I told the code to use the new "standard" connection path (standard-ftms) based on Horizon 7.0 AT, it started working perfectly.

Right now, I have set the code to use the Wahoo path by default. If a user wants the Standard FTMS, they have to use a switch in the command line.

How would you like to handle this? Is this approach ok?

cagnulein commented:
Thanks @Bassai-Sho for your massive work!
I will start a review soon!

For the wahoo tread part: if you simply change the name to something different from the wahoo name, was it still crashing? was the runna app crashes also when you run qz from your mobile?

Bassai-Sho commented:
here is the response from Claude via Vscode

The root cause was that Runna detected the Wahoo proprietary GATT service (A026EE0E-...) advertised by the virtual treadmill and switched to the Wahoo BLE protocol instead of standard FTMS. The Wahoo protocol skips subscribing to the FTMS Control Point CCCD, which caused a 6-second timeout and immediate disconnect (Crash).

The fix was removing that Wahoo service from the BLE peripheral entirely when running in standard FTMS mode (via -no-wahoo-service). Without it, Runna falls back to standard FTMS, correctly subscribes to the Control Point, and the connection holds. Everything else — the device name, DIS strings, feature flags — was supporting work to make the device look like a known standard FTMS treadmill so Runna trusted it.

I did not check with two android devices as i only have the one, but I can check over the weekend (heading into work now)

cagnulein commented:
so i guess the best way to approach this is to create a new setting to enable/disable the wahoo tread (maybe called Runna or similar). What do you think? I would like to keep the ANT_ defines at less as possible

Bassai-Sho commented:
Agree, that was my thought , the guards were made to be easily removed, given this feels not just limited to PI. The setting would need to be in the QML gui, but not sure how to go about this, can certainly attempt, however this Runna issue seems more of a higher priority. I can raise this as a new branch to just get Runna working then look for some help to get the GUI settings made? Unless you are happy to keep as is and when I fix the ant+ setup-dashboard issue we do this as one update?

cagnulein commented:
if you want i can create the setting and the UI for you so you can link in the virtualtreadmill.cpp, what do you think?

Bassai-Sho commented:
That would be awesome, thank you 😊. Then I just point to that setting and remove the no-wahoo switch.

cagnulein committed:
wahoo virtual treadmill setting

cagnulein commented:
done, it's under the treadmill options, enabled by default

Bassai-Sho committed:
fix: reconnect FTMS/Wahoo signals after BLE disconnect

Qt signals on serviceFTMS and serviceWahoo were only wired in the
constructor. reconnect() creates new service instances via addService(),
leaving the constructor connections pointing at dead objects. Slope
commands from Runna were silently dropped on every reconnect, causing
a crash.

Fix: re-connect characteristicChanged signals in reconnect() after
each addService() call.

Remove wahoo_treadmill_emulation setting and all associated UI/config
code — it was an unnecessary workaround chasing the wrong root cause.

Bassai-Sho committed:
Merge branch 'master' into feat/ant_footpod

Bassai-Sho committed:
fix(dashboard): remove wahoo emulation setting and fix stray terminal output

Bassai-Sho committed:
docs: add Known Issues section and inline callouts to README

Bassai-Sho commented:
Hi Roberto, so the Runna crashing was actually more simpler than first thought, it was resolved by adding the serviceFTMS and serviceWahoo signals to the reconnect(). Without the reconnect they were pointing to dead objects. No need for the new flags or additional workaround, which is better overall. Have tested and works.

There are some known issues but nothing that prevents the core functionality from working:
- Virtual Bluetooth devices (e.g. Just Fit) do not work on BlueZ 5.70+ / Ubuntu 24.04 — ANT+ footpod unaffected
- SD card protection (Overlay FS) is not currently working and needs investigation — do not rely on it
- ANT+ footpod flag will default to false  this needs to be enabled via QZ Runtime Flags and service regenerated
- Device online/offline status does not always reflect the true current state
- Guided setup wizard does not offer Smart Monitor — must be set up manually after wizard completes
- Occasional screen artefacts in the dashboard — display only, does not affect functionality

I can look at these later, but thought it is ready enough especially with the Runna fix. :+1:

cagnulein commented:
> Hi Roberto, so the Runna crashing was actually more simpler than first thought, it was resolved by adding the serviceFTMS and serviceWahoo signals to the reconnect(). Without the reconnect they were pointing to dead objects. No need for the new flags or additional workaround, which is better overall. Have tested and works.

That's an awesome news! Let me review it and I will merge it then! thanks!

cagnulein commented:
hi @Bassai-Sho i was reviewing it, were you able to test this also with zwift? it's still working? i was worried about https://github.com/cagnulein/qdomyos-zwift/blob/3e3149ce311edc51c4320d6f683f0bd91bf3f24f/src/virtualdevices/virtualtreadmill.cpp#L147-L155

Thanks!

Bassai-Sho commented:
HI @cagnulein,

I have not used Zwift for while now. however here is my reasoning and steps followed so far.

- I just did a test using Zwift on my Android phone, selecting "Kicker Run" for both [Run Speed] and [Controllable] tiles in Zwift.
- The assumption is, Incline will automatically adjust.
- I choose Hill + Tempo # 3, and 12 minutes into the run it was asking to change inclination to ~4%. The  treadmill did not automatically adjust the incline.
- The speed and incline from treadmill was succesfully getting pushed to Zwift
- The ANT+ metrics were succesfully being sent to the watch.

So your concern around incline might be valid , I will look to do the following additional steps

- Create a custom .TWO workout, which will quickly adjust the inclination
- Use two Android devices , Android 1 for  QZ and Android 2 for Zwift, to validate inclination adjusts automatically prior to this change.

My initial thinking is the code segment might require gating if Zwift is being used, still allowing 3rd party apps such as Runna to work. But I can experiement and let you know how I get on.

Bassai-Sho commented:
Hi @cagnulein,

Ok, the rasperry pi **worked**!!  The two Android setup also worked as a baseline validation.

I followed https://github.com/cagnulein/qdomyos-zwift/wiki/Can-Zwift-control-speed-and-or-inclination-of-my-treadmill%3F

I needed to have been in a **Free Run** in Zwift  (running on android). I selected "Mountain Mash" and travelled ~1km into the run. The incline on the treadmill started to self adjust. Seems that Workouts do not send gradient information due to ERG/pace mode, where Free runs use the actual terrian and gradient data.

I also learnt that custom run workouts using .ZWO files cannot include gradient/incline information so stopped that idea. https://www.zwiftworkout.com/

I attach the debug file as evidence
[debug-Thu_May_28_16_02_24_2026.log](https://github.com/user-attachments/files/28352716/debug-Thu_May_28_16_02_24_2026.log)

Hope this helps :+1:
